### PR TITLE
Locked timeout

### PR DESCRIPTION
**Pull request by bstanescu**
_Spinlock with a twist (pun intended 😉)_

**Problem:**
I needed to throw an exception if the lock is already taken, but only after a given timeout. 

**Proposed solution**
The spinlock already has its own timeout (eg: 5 seconds). With this new option: `lockedTimeout` you can now specify a timeout while lock is already taken. (eg: 3 seconds).
The process will try to acquire the lock for 3 seconds then throw an exception if it doesn't succeed.

If we specify a `lockedTimeout` of 0, the exception will be thrown immediately after the first lock acquire attempt.

This (with a `lockedTimeout` of zero) also addresses:
https://github.com/php-lock/lock/issues/21#issuecomment-394963109

### CODE REVIEW

File: classes/exception/LockedTimeoutException.php [hunk 1]
```diff
@@ -0,0 +1,10 @@
+<?php
+
```
Comment (TheLevti) [inline classes/exception/LockedTimeoutException.php:2]:
```suggestion

declare(strict_types=1);

```

**Comment by willemstuursma** [comment]
Thanks. So you want to have a different timeout for acquiring the lock and the maximum lock execution, correct?

I think we should name the two locks then `$acquisitionTimeout` and `$executionTimeout`.

**Comment by willemstuursma** [comment]
Could you update all the files to change the naming, I'll have a better look in the coming days.

File: classes/mutex/RedisMutex.php [hunk 2]
```diff
@@ -42,12 +42,13 @@ abstract class RedisMutex extends SpinlockMutex implements LoggerAwareInterface
      * @param array  $redisAPIs The Redis APIs.
      * @param string $name      The lock name.
      * @param int    $timeout   The time in seconds a lock expires, default is 3.
+     * @param int    $lockedTimeout The maximum time in seconds we are spinning while locked.
      *
      * @throws \LengthException The timeout must be greater than 0.
      */
-    public function __construct(array $redisAPIs, string $name, int $timeout = 3)
+    public function __construct(array $redisAPIs, string $name, int $timeout = 3, int $lockedTimeout = null)
```
Comment (TheLevti) [inline classes/mutex/RedisMutex.php:49]: `?int`

**Comment by mvorisek** [comment]
@bstanescu please rebase this PR on the latest `master`

**Comment by mvorisek** [comment]
There is newly https://github.com/php-lock/lock/blob/c9ad6c1655/src/Mutex/AbstractSpinlockWithTokenMutex.php#L61 with "expire timeout" ("locked timeout" used here) - let's add this condition to the `AbstractSpinlockWithTokenMutex::acquire()` method and have this PR merged.

**Pull request closed without merging**